qiskit/algorithms/optimizers/spsa.py

```diff
@@ -49,195 +53,335 @@ class SPSA(Optimizer):
+        # compute the gradient approximation and additionally return the loss function evaluations
+        plus, minus = loss(x + eps * delta), loss(x - eps * delta)
+        gradient_sample = (plus - minus) / (2 * eps) * delta
+        self._nfev += 2
+
+        return gradient_sample
+
+    def _point_estimate(self, loss, x, eps, deltas):
+        """The gradient estimate at point ``x`` consisting as average of all directions ``delta``.
+        """
+        # number of samples
+        resamplings = len(deltas)
+
+        # set up variables to store averages
+        gradient_estimate = np.zeros(x.size)
+
+        # iterate over the directions
+        for delta in deltas:
+            gradient_sample = self._point_sample(loss, x, eps, delta)
+            gradient_estimate += gradient_sample
+
+        return gradient_estimate / resamplings
+
+    def _compute_update(self, loss, x, k, eps):
+        # compute the perturbations
+        if isinstance(self.resamplings, dict):
+            avg = self.resamplings.get(k, 1)
+        else:
+            avg = self.resamplings
+
+        gradient = np.zeros(x.size)
+
+        # accumulate the number of samples
+        deltas = [bernoulli_perturbation(x.size, self.perturbation_dims) for _ in range(avg)]
+
+        gradient = self._point_estimate(loss, x, eps, deltas)
+
+        return gradient
+
+    def _minimize(self, loss, initial_point):
+        # ensure learning rate and perturbation are set
+        # this happens only here because for the calibration the loss function is required
+        if self.learning_rate is None and self.perturbation is None:
+            get_learning_rate, get_perturbation = self.calibrate(loss, initial_point)
+            self.learning_rate = get_learning_rate
+            self.perturbation = get_perturbation
+
+        if self.learning_rate is None or self.perturbation is None:
+            raise ValueError('If one of learning rate or perturbation is set, both must be set.')
+
+        # get iterator
+        eta = self.learning_rate()
+        eps = self.perturbation()
+
+        # prepare some initials
+        x = np.asarray(initial_point)
+
+        self._nfev = 0
+
+        # if blocking is enabled we need to keep track of the function values
+        if self.blocking:
+            fx = loss(x)
+
+            self._nfev += 1
+            if self.allowed_increase is None:
+                self.allowed_increase = 2 * self.estimate_stddev(loss, x)
+
+        logger.info('=' * 30)
+        logger.info('Starting SPSA optimization')
+        start = time()
+
+        # keep track of the last few steps to return their average
+        last_steps = deque([x])
+
+        for k in range(1, self.maxiter + 1):
+            iteration_start = time()
+            # compute update
+            update = self._compute_update(loss, x, k, next(eps))
+
+            # trust region
+            if self.trust_region:
```

Comment:
why do we need this?  Doesn't the `c` parameter in the alg already control the maximum step that can be taken?

Reply:
one is rejecting steps that go to large, do we not want big downward changes to be accepted, i.e. don't we want to follow steep gradients downward?

Reply:
That can be achieved using the blocking condition (`blocking=True`), this attribute is really just to confine the size of the update step to improve stability in possibly too noisy functions.